daemon/mgr/container_diskquota.go

```diff
@@ -0,0 +1,117 @@
+package mgr
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/alibaba/pouch/pkg/quota"
+	"github.com/docker/docker/pkg/reexec"
+	"github.com/sirupsen/logrus"
+)
+
+func init() {
+	reexec.Register("set-diskquota", processSetQuotaReexec)
+}
+
+func processSetQuotaReexec() {
+	var (
+		err      error
+		upperDir string
+	)
+
+	// Return a failure to the calling process via ExitCode
+	defer func() {
+		if err != nil {
+			logrus.Fatalf("%v", err)
+		}
+	}()
+
+	if len(os.Args) != 3 {
+		err = fmt.Errorf("command args is invalid: %v", os.Args)
+		return
+	}
+
+	basefs := os.Args[1]
+	size := os.Args[2]
+
+	logrus.Infof("set diskquota: %v", os.Args)
+
+	upperDir, err = getOverlay2UpperDir(basefs)
+	if err != nil || upperDir == "" {
+		logrus.Errorf("failed to get lowerdir: %v", err)
+		return
+	}
+
+	_, err = quota.StartQuotaDriver(upperDir)
+	if err != nil {
+		logrus.Errorf("failed to start quota driver: %v", err)
+		return
+	}
+
+	qid, err := quota.SetSubtree(upperDir, 0)
+	if err != nil {
+		logrus.Errorf("failed to set subtree: %v", err)
+		return
+	}
+
+	err = quota.SetDiskQuota(upperDir, size, int(qid))
+	if err != nil {
+		logrus.Errorf("failed to set disk quota: %v", err)
+	}
+
+	// FIXME: long time for start container.
+	//go setQuotaForDir(upperDir, qid)
+
+	return
+}
+
+func setQuotaForDir(src string, updateQid uint32) {
+	filepath.Walk(src, func(path string, fd os.FileInfo, err error) error {
+		if err != nil {
+			logrus.Warnf("setQuota walk dir %s get error %v", path, err)
+			return nil
+		}
+		quota.SetFileAttrNoOutput(path, updateQid)
+		return nil
+	})
+}
+
+func getOverlay2UpperDir(basefs string) (string, error) {
+	var upperDir string
+
+	fd, err := os.Open("/proc/mounts")
+	if err != nil {
+		return "", err
+	}
+	defer fd.Close()
+
+	br := bufio.NewReader(fd)
+	for {
+		line, _, c := br.ReadLine()
+		if c == io.EOF {
+			break
+		}
+
+		parts := strings.Split(string(line), " ")
+		if len(parts) != 6 {
+			continue
+		}
+		if parts[1] != basefs || parts[2] != "overlay" {
+			continue
+		}
+
+		mountParams := strings.Split(parts[3], ",")
+		for _, p := range mountParams {
+			if strings.Contains(p, "upperdir") {
+				upperDir = strings.Split(p, "=")[1]
```

Comment:
Does this code have the potential risk to panic?
Directly use the second element [1] in a slice?